python_modules/dagster/dagster/_core/definitions/asset_job.py

```diff
@@ -295,8 +300,20 @@ def _subset_assets_defs(
             selected_check_subset = selected_asset_check_keys & asset.check_keys
         # if no checks were selected, filter to checks that target selected assets
         else:
+            print("I GOT HIT")
+            # In the very specific case of dbt sources; upstream stub assets may have checks defined in the assets definition.
+            # In this case, we want to include the checks that target the upstream assets.
+            upstreams = {
+                dep.asset_key
+                for spec in asset.specs
+                for dep in spec.deps
+                if spec.key in selected_asset_keys
+            }
+            asset_keys_that_checks_can_target = selected_asset_keys | upstreams
```

Comment:
the specific case that I don't like here is that this captures upstreams that are not dbt sources in the dbt case.

e.g. you have source -> model_a -> model_b, now executing just model_b will execute any asset checks defined on model_a.

I think we need an additional filter to only select checks for assets that are not defined on the asset at all (and are upstreams of selected assets).

That logic feels a bit arbitrary and perhaps over-fit, but the real original sin here is the fact that asset_check_selection can be None, and so at this point we gotta just create some sort of least-surprising behavior.

The other-other option here is to just leave this logic untouched now that we warn instead of error on emitting unexpected asset check results. Thoughts on that instead?